compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmBuiltinOptimizationLowering.kt

```diff
@@ -93,6 +92,69 @@ class JvmBuiltinOptimizationLowering(val context: JvmBackendContext) : FileLower
                     expression
                 }
             }
+
+            override fun visitWhen(expression: IrWhen): IrExpression {
+                expression.transformChildrenVoid(this)
+                // Remove all branches with constant false condition.
+                expression.branches.removeIf() {
+                    it.condition.isFalseConst()
+                }
+                // If the only condition that is left has a constant true condition remove the
+                // when in favor of the result. If there are no conditions left, remove the when
+                // entirely and replace it with an empty block.
+                if (expression.branches.size == 1) {
+                    val branch = expression.branches.first()
+                    if (branch.condition.isTrueConst()) return branch.result
+                }
+                return if (expression.branches.size == 0)
+                    IrBlockImpl(expression.startOffset, expression.endOffset, context.irBuiltIns.unitType)
+                else
+                    expression
+            }
+
+            private fun isImmutableTemporaryVariableWithConstantValue(statement: IrStatement): Boolean {
+                return statement is IrVariable &&
+                        statement.descriptor is IrTemporaryVariableDescriptor &&
```

Comment:
Please avoid descriptors in ir lowerings when possible, we are still in process of removing them from ir transformations.
Seems next code would be equivalent: statement.origin == IrDeclarationOrigin.IR_TEMPORARY_VARIABLE

Reply:
Done, thanks!